### PR TITLE
GH-865: NFAR removal

### LINKED ISSUE #865: As a smith I want to remove N4 archive completely
N4 archives / nfars are not used in production. New features (e.g. package.json and others) would be adjusted to nfars or vice versa, nfars might have to be adjusted to the new features. Since nfars are not used, it is reasonable to remove them completely from the code base.

### PR DESCRIPTION
**Pull request by lbeurerkellner**
Fixes #865 (actually a userstory)

With this PR, I have removed all remaining support for NFAR archives. This include, among others, the following points:

* Removed `IN4JSArchive` + subclasses and corresponding special handling in the N4JS project model (this includes `getLibraries` and related methods)
* Removed `IN4JSSourceContainerAware` abstraction. The only source-container-aware entity that remains is `IN4JSProject`
* Updated spec (removed paragraph on archives)

### CODE REVIEW

File: plugins/org.eclipse.n4js.runner/src/org/eclipse/n4js/runner/RuntimeEnvironmentsHelper.java [hunk 1]
```diff
@@ -123,9 +121,9 @@ private void getEnvironemntWithAncestorsRecursive(IN4JSProject project, Set<IN4J
 		if (!project.getProjectType().equals(ProjectType.RUNTIME_ENVIRONMENT)) {
 			return;
 		}
-		Optional<IN4JSSourceContainerAware> maybePArent = project.getExtendedRuntimeEnvironment();
+		Optional<IN4JSProject> maybePArent = project.getExtendedRuntimeEnvironment();
```
Comment (yoosiba) [inline plugins/org.eclipse.n4js.runner/src/org/eclipse/n4js/runner/RuntimeEnvironmentsHelper.java]: Pls fix my typos! `maybePArent` looks like something I would commit :expressionless:

File: plugins/org.eclipse.n4js.runner/src/org/eclipse/n4js/runner/RunnerHelper.java [hunk 2]
```diff
@@ -197,7 +195,7 @@ public ApiUsage getProjectExtendedDeps(String runnerId, URI moduleToRun) {
 	 *
 	 * @see RuntimeEnvironmentsHelper#recursiveDependencyCollector
 	 */
-	public void recursiveExtendedREsCollector(IN4JSSourceContainerAware sourceContainer,
+	public void recursiveExtendedREsCollector(IN4JSProject sourceContainer,
```
Comment (yoosiba) [inline plugins/org.eclipse.n4js.runner/src/org/eclipse/n4js/runner/RunnerHelper.java]: in some places parameter is renamed. Rename for consistency?

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/internal/N4MFProjectDependencyStrategy.java [hunk 3]
```diff
@@ -64,7 +64,7 @@
 					.filter(IN4JSEclipseProject.class).toList();
 
 			// This is for the extended RE, if specified.
-			final IN4JSSourceContainerAware srcContainerAware = n4Project.getExtendedRuntimeEnvironment().orNull();
+			final IN4JSProject srcContainerAware = n4Project.getExtendedRuntimeEnvironment().orNull();
```
Comment (yoosiba) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/internal/N4MFProjectDependencyStrategy.java]: rename as in other places?

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/helper/SourceContainerAwareDependencyTraverser.xtend [hunk 4]
```diff
@@ -17,31 +17,30 @@ import com.google.common.collect.ImmutableList
...
-import org.eclipse.n4js.projectModel.IN4JSSourceContainerAware
 import org.eclipse.n4js.utils.DependencyTraverser
 
 /**
- * Class for traversing {@link IN4JSSourceContainerAware source container aware} dependencies and
+ * Class for traversing {@link IN4JSProject source container aware} dependencies and
  * indicating cycles in the dependency graph.
  */
-class SourceContainerAwareDependencyTraverser extends DependencyTraverser<IN4JSSourceContainerAware> {
+class SourceContainerAwareDependencyTraverser extends DependencyTraverser<IN4JSProject> {
```
Comment (yoosiba) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/helper/SourceContainerAwareDependencyTraverser.xtend]: Can it be named `ProjectDependencyTraverser`?
Do we still need generic `DependencyTraverser`?

**Comment by lbeurerkellner** [comment]
The merge of this PR is postponed until the first clean-up by @mor-n4 of GH-809 (package.json) is merged.

File: plugins/org.eclipse.n4js.npmexporter/src/org/eclipse/n4js/npmexporter/ConvertManifestToPackageJson.xtend [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+import org.eclipse.n4js.runner.^extension.RuntimeEnvironment
+import java.nio.file.Path
+
+import org.eclipse.n4js.semver.model.SemverSerializer
+
+// GH-809 delete this class
+/**
+ * Converting all relevant Manifest-data into an DataObject.
+ */
+class ConvertManifestToPackageJson {
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js.npmexporter/src/org/eclipse/n4js/npmexporter/ConvertManifestToPackageJson.xtend]: This class was deleted as part of the `GH-809` clean up. Now it seems you re-create this class.

File: plugins/org.eclipse.n4js.runner/src/org/eclipse/n4js/runner/RuntimeEnvironmentsHelper.java [hunk 6]
```diff
@@ -137,23 +134,23 @@
...
-		final DependencyCycle<IN4JSSourceContainerAware> result = traverser.findCycle();
+		final DependencyCycle<IN4JSProject> result = traverser.findCycle();
 
 		// check whether a dependency cycle between workspace projects was discovered
 		if (result.hasCycle()) {
 			throw new DependencyCycleDetectedException(project);
 		}
 
-		return ImmutableList.copyOf(runtimeLibraryDependencies);
+		return runtimeLibraryDependencies;
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js.runner/src/org/eclipse/n4js/runner/RuntimeEnvironmentsHelper.java]: why change the return value from immutable to mutable?

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/containers/N4JSToBeBuiltComputer.java [hunk 7]
```diff
@@ -60,28 +55,9 @@
...
-		storageMapper.collectNfarURIs(project, toBeBuilt.getToBeDeleted());
-		Iterator<Map.Entry<String, Set<URI>>> iterator = knownEntries.entrySet().iterator();
-		String keyPrefix = project.getName() + "|";
-		while (iterator.hasNext()) {
-			String key = iterator.next().getKey();
-			if (key.startsWith(keyPrefix)) {
-				iterator.remove();
-			}
-		}
+		// nothing to remove
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/containers/N4JSToBeBuiltComputer.java]: do we even need this method, then? (same for `#removeStorage()` below)

Comment (lbeurerkellner) [reply plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/containers/N4JSToBeBuiltComputer.java]: It is an interface implementation, thus we need the stubs (cf. `IToBeBuiltComputerContribution`).

**Pull request merged**